src/main/java/edu/cornell/kfs/ksr/web/struts/SecurityRequestDocumentAction.java

```diff
@@ -0,0 +1,280 @@
+
+            String qualificationIndexStr = StringUtils.substringBetween(parameterName, ".qualificationIndex", ".");
+            qualificationIndex = Integer.parseInt(qualificationIndexStr);
+        } else {
+            LOG.error("deleteQualificationLine() Unable to find qualification index for line to delete");
+            throw new RuntimeException("Unable to find qualification index for line to delete");
+        }
+
+        SecurityRequestRole securityRequestRole = document.getSecurityRequestRoles().get(roleRequestIndex);
+        if (securityRequestRole == null) {
+            LOG.error("deleteQualificationLine() Security request role not found for index: " + roleRequestIndex);
+            throw new RuntimeException("Security request role not found for index: " + roleRequestIndex);
+        }
+
+        securityRequestRole.getRequestRoleQualifications().remove(qualificationIndex);
+
+        return mapping.findForward(KFSConstants.MAPPING_BASIC);
+    }
+
+    @Override
+    public ActionForward refresh(ActionMapping mapping, ActionForm form, HttpServletRequest request, HttpServletResponse response) throws Exception {
+        SecurityRequestDocumentForm documentForm = (SecurityRequestDocumentForm) form;
+        SecurityRequestDocument document = (SecurityRequestDocument) documentForm.getDocument();
+
+        super.refresh(mapping, form, request, response);
+
+        if (StringUtils.isNotBlank(request.getParameter("document.principalId"))) {
+
+            if (StringUtils.isNotBlank(document.getPrincipalId()) && ObjectUtils.isNotNull(document.getRequestPerson())) {
+                if (!StringUtils.equals(document.getPrincipalId(), documentForm.getCurrentPrincipalId())) {
+                    SpringContext.getBean(SecurityRequestDocumentService.class).initiateSecurityRequestDocument(document,
+                            GlobalVariables.getUserSession().getPerson());
+                    documentForm.setTabRoleIndexes(buildTabRoleIndexes(document));
+                    documentForm.setCurrentPrincipalId(document.getPrincipalId());
+                }
+            }
+        }
+
+        return mapping.findForward(KFSConstants.MAPPING_BASIC);
+    }
+
+    protected String getReturnLocation(HttpServletRequest request, ActionMapping mapping) {
+        return getApplicationBaseUrl() + mapping.getPath() + ".do";
+    }
+
+    @SuppressWarnings({ "rawtypes", "unchecked" })
+    protected List<SecurityRequestDocumentForm.TabRoleIndexes> buildTabRoleIndexes(final SecurityRequestDocument document) {
+        final List<SecurityRequestDocumentForm.TabRoleIndexes> tabRoleIndexes = new ArrayList<SecurityRequestDocumentForm.TabRoleIndexes>();
+
+        List sortPropertyNames = new ArrayList();
+        sortPropertyNames.add(KSRPropertyConstants.SECURITY_REQUEST_DOCUMENT_TAB_ORDER);
+        
+        List<SecurityGroupTab> securityGroupTabs = document.getSecurityGroup().getSecurityGroupTabs().stream().collect(Collectors.toList());

```

Comment:
On further consideration, I recall that the Java 8 Javadocs mention that Collectors.toList() is not guaranteed to return a mutable list. It might be safer to rewrite that as "Collectors.toCollection(ArrayList::new)" instead.

Sorry for not noticing this earlier.

Reply:
No problem, thank you very much for catching this. Will update.